### PR TITLE
Schedule reroute after allocator timed out

### LINKED ISSUE #14945: [BUG] Schedule reroute after allocator timeout
### Describe the bug

When the existing shards allocator would timeout in batch mode, we would need to attempt another reroute such that next iteration is in line. Currently, applyStartedShards, applyFailedShards, async fetch, etc would internally trigger a subsequent reroute from the previous batch. But this behaviour might not be guaranteed.

Related PR comment - https://github.com/opensearch-project/OpenSearch/pull/14848#discussion_r1689072743

### Related component

Other

### Expected behavior

Trigger subsequent reroute on timeout

### Additional Details

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by imRishN**
### Description
This PR attempts to schedule a reroute whenever an allocator times out. This would ensure that we never let a shard in an unassigned state because the current reroute cycle timed out.

### Related Issues
Resolves #14945 

### Check List
- [ ] Functionality includes testing.
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/ClusterModule.java [hunk 1]
```diff
@@ -474,4 +475,9 @@ public void setExistingShardsAllocators(GatewayAllocator gatewayAllocator, Shard
         allocationService.setExistingShardsAllocators(existingShardsAllocators);
     }
 
+    public void setRerouteServiceForAllocator(RerouteService rerouteService) {
+        if (shardsAllocator instanceof BalancedShardsAllocator) {
+            ((BalancedShardsAllocator) shardsAllocator).setRerouteService(rerouteService);
+        }
+    }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/ClusterModule.java:482]: Can we have a default implementation in ShardAllocator and override the behaviour in BSA?

Comment (imRishN) [reply server/src/main/java/org/opensearch/cluster/ClusterModule.java]: Made changes. Can you check if this looks good?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 2]
```diff
@@ -404,6 +415,23 @@ private void failAllocationOfNewPrimaries(RoutingAllocation allocation) {
         }
     }
 
+    private void scheduleRerouteIfAllocatorTimedOut() {
+        if (allocatorTimedOut()) {
+            if (rerouteService == null) {
+                logger.info("RerouteService not set to schedule reroute after allocator time out");
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:421]: What does this log denote? I would rather be more assertive if this was expected to be null or not?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 3]
```diff
@@ -231,6 +235,12 @@ public BalancedShardsAllocator(Settings settings, ClusterSettings clusterSetting
         clusterSettings.addSettingsUpdateConsumer(ALLOCATOR_TIMEOUT_SETTING, this::setAllocatorTimeout);
     }
 
+    @Override
+    public void setRerouteService(RerouteService rerouteService) {
+        assert this.rerouteService == null : "RerouteService is already set";
+        this.rerouteService = rerouteService;
+    }
+
```
[the remark this answers is not in the document]
Comment (imRishN) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:243]: For BSA, during node construction, reroute service is initialised. This assert just pre-emptively checks that reroute service is just set once. Making it optional is not really needed. Wdyt?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java (hunk 2 above)
Comment (imRishN) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: Updated. Added initially for tests

File: server/src/main/java/org/opensearch/cluster/ClusterModule.java [hunk 4]
```diff
@@ -474,4 +475,9 @@ public void setExistingShardsAllocators(GatewayAllocator gatewayAllocator, Shard
         allocationService.setExistingShardsAllocators(existingShardsAllocators);
     }
 
+    public void setRerouteServiceForAllocator(RerouteService rerouteService) {
+        if (shardsAllocator instanceof BalancedShardsAllocator) {
+            shardsAllocator.setRerouteService(rerouteService);
+        }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/ClusterModule.java:481]: we don't need the check here, the classes can themselves based on the overridden methods handle it?

File: server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java [hunk 5]
```diff
@@ -198,6 +198,17 @@ protected ShardsBatchGatewayAllocator(long batchSize) {
...
+        this.rerouteService = rerouteService;
+        this.batchStartedAction = null;
+        this.primaryShardBatchAllocator = null;
+        this.batchStoreAction = null;
+        this.replicaShardBatchAllocator = null;
+        this.maxBatchSize = DEFAULT_SHARD_BATCH_SIZE;
+        this.primaryShardsBatchGatewayAllocatorTimeout = null;
+        this.replicaShardsBatchGatewayAllocatorTimeout = null;
+    }
+
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java:211]: Can we create a common ctor 
```
protected ShardsBatchGatewayAllocator(RerouteService rerouteService, long batchSize) {
        this.rerouteService = rerouteService;
        this.batchStartedAction = null;
        this.primaryShardBatchAllocator = null;
        this.batchStoreAction = null;
        this.replicaShardBatchAllocator = null;
        this.maxBatchSize = batchsize;
        this.primaryShardsBatchGatewayAllocatorTimeout = null;
        this.replicaShardsBatchGatewayAllocatorTimeout = null;
    }
```

and the current ctor just do this(blah, blah...) and invoke the common ctor

Comment (imRishN) [reply server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java]: Updated

**Review by Bukhtawar — APPROVED** [review]

**Comment by imRishN** [comment]
Successful build - https://build.ci.opensearch.org/job/gradle-check/46694/

**Comment by imRishN** [comment]
Build failure because of 1 flaky test - https://build.ci.opensearch.org/job/gradle-check/46860/

**Review by Bukhtawar — APPROVED** [review]

**Pull request merged**